### ISSUE TITLE
Suggestion: Only keep the most recent patch information in opcodes.jsonc

### ISSUE THREAD

**Issue by Makar8000** [OverlayPlugin/OverlayPlugin#235]
I'd like to propose the idea of only keeping the most recent patch in the [`opcodes.jsonc`](https://github.com/OverlayPlugin/OverlayPlugin/blob/main/OverlayPlugin.Core/resources/opcodes.jsonc) file. Example:
```jsonc
{
    // Region: Global
    // Version: 6.41
    "2023.06.14.0000.0000": {
        "MapEffect": {
            "opcode": 736,
            "size": 11
        },
        "CEDirector": {
            "opcode": 392,
            "size": 16
        },
        "RSVData": {
            "opcode": 142,
            "size": 1080
        }
    },
    // Region: CN
    // Version: 6.35
    "2023.06.28.0000.0000": {
        "MapEffect": {
            "opcode": 782,
            "size": 11
        },
        "CEDirector": {
            "opcode": 998,
            "size": 16
        },
        "RSVData": {
            "opcode": 783,
            "size": 1080
        }
    },
    // Region: KR
    // Version: 6.3
    "2023.06.02.0000.0000": {
        "MapEffect": {
            "opcode": 629,
            "size": 11
        },
        "CEDirector": {
            "opcode": 777,
            "size": 16
        },
        "RSVData": {
            "opcode": 115,
            "size": 1080
        }
    },
}
``` 

Why?
- GitHub itself will maintain history of past patches if it's necessary.
- It's not possible to run any game client versions other than latest with retail servers, and the state of things like Sapphire is quite far from stable/combat ready.
- Only maintaining one kv per region helps prevent merge conflicts.
  - A merge conflict has [happened before](https://github.com/OverlayPlugin/OverlayPlugin/pull/223) and will likely happen again with 6.45 retail and 6.31 KR happening at the same time.
  - Merge conflicts will potentially cause delays for users given that a beta version of the parser plugin is likely to release quickly for opcode-only updates (most minor patches).
 - Reduce filesize of the payload when fetching opcodes remotely (very minor, but figured I would mention it anyway).

Basically, I'm not sure of the initial reasons for opting to maintain history when updating the `opcodes.jsonc` file, and I wonder if those reasons are still applicable today. As a side note, I'm also curious of the possibility of the versions being the same between two regions... maybe the answer would be to make the key the region rather than patch ver.

**Comment by valarnin** [comment]
The original intent of the file wasn't to keep every version of opcode, but rather to allow for quick lookup of version opcode info without having a conflicting key. Realistically, if JSON allowed for a compound key, the key would have just been region+version to begin with.

Basically, there were two design requirements:

1. The opcode definition must include the version, to prevent loading the wrong opcode for the current game version
2. All lookups must be quick, as the lookup happens on packet received and must be resolved quickly to prevent locking threads

> As a side note, I'm also curious of the possibility of the versions being the same between two regions... maybe the answer would be to make the key the region rather than patch ver.

- Game versions are the build date of the executable, not the date that the patch drops. That makes it exceedingly unlikely that there is a version conflict.
- The reason that the region wasn't used as the key was to avoid the extra string compare while processing packet data

**Comment by Makar8000** [comment]
> Game versions are the build date of the executable, not the date that the patch drops. That makes it exceedingly unlikely that there is a version conflict.

I was aware that game versions were the build date; it didn't really change how I felt about it being "not impossible" - we've had situations where it was already pretty close, so I was just curious.

> The reason that the region wasn't used as the key was to avoid the extra string compare while processing packet data

Makes sense.

> The original intent of the file wasn't to keep every version of opcode, but rather to allow for quick lookup of version opcode info without having a conflicting key. Realistically, if JSON allowed for a compound key, the key would have just been region+version to begin with.
> 
> Basically, there were two design requirements:
> 
> 1. The opcode definition must include the version, to prevent loading the wrong opcode for the current game version
> 2. All lookups must be quick, as the lookup happens on packet received and must be resolved quickly to prevent locking threads

So, if I'm understanding this correctly - the intent was never to keep "adding onto" the file, and we can remove the old versions? If so, I can make a quick PR to strip it, so that we can more likely avoid the merge conflicts 😅
